lib/semantics/check-declarations.cc

```diff
@@ -343,6 +347,18 @@ void CheckHelper::CheckObjectEntity(
       }
     }
   }
+  if (symbol.owner().kind() != Scope::Kind::DerivedType &&
+      IsInitialized(symbol)) {
+    if (details.commonBlock()) {
+      if (details.commonBlock()->name().empty()) {
+        messages_.Say(
+            "A variable in blank COMMON should not be initialized"_en_US);
+      }
+    } else if (symbol.owner().kind() == Scope::Kind::BlockData) {
```

Comment:
Is the error message that the compiler prints important?

When I run this test:
block data test
  integer, allocatable :: t1
end block data
It prints the message:
An initialized variable in BLOCK DATA must be in a COMMON block

The compiler is behaving correctly by printing an error message. 
But is this an initialization of the variable? I guess you could say yes because it dynamically allocates space for the array.
But allocatable is also in the clause C1416. 
Gfortran prints:
ALLOCATABLE attribute not allowed in BLOCK DATA program unit at
Is that a way to make the message more clear?

With the current message the user would try to solve by adding in the common, and it will not work because allocatable is not allowed in common.

Reply:
`ALLOCATABLE` variables are initialized to the unallocated state.

I will improve the message.

Reply:
Thanks.